lib/stat_tracker.rb

```diff
@@ -14,34 +21,314 @@ def self.from_csv(data_files)
+      game.home_team_id == team_id
+    end
+  end
+
+  def home_games_grouped_by_season(team_id)
+    home_games_filtered_by_team(team_id).group_by do |game|
+      game.season
+    end
+  end
+
+  def season_home_wins(team_id)
+    wins = Hash.new(0.0)
+    home_games_grouped_by_season(team_id).each do |season, game|
+      game.each do |game|
+        if game.outcome == :home_win
+          wins[season] += 1.0
+        elsif game.outcome == :away_win
+          wins[season] -= 1.0
+        else
+          wins[season] -= 0.5
+        end
+      end
+    end
+    wins
+  end
+
+  def away_games_filtered_by_team(team_id)
+    game_collection.all.find_all do |game|
+      game.away_team_id == team_id
+    end
+  end
+
+  def away_games_grouped_by_season(team_id)
+    away_games_filtered_by_team(team_id).group_by do |game|
+      game.season
+    end
+  end
+
+  def season_away_wins(team_id)
+    wins = Hash.new(0.0)
+    away_games_grouped_by_season(team_id).each do |season, game|
+      game.each do |game|
+        if game.outcome == :away_win
+          wins[season] += 1.0
+        elsif game.outcome == :home_win
+          wins[season] -= 1.0
+        else
+          wins[season] -= 0.5
+        end
+      end
+    end
+    wins
+  end
+
+  def win_count_by_season(team_id)
+    accum = {}
+    season_away_wins(team_id).each do |season_away_id, away_win|
+      season_home_wins(team_id).each do |season_home_id, home_win|
+        if season_away_id == season_home_id
+          accum[season_away_id] = away_win + home_win
+        end
+      end
+    end
+    accum
+  end
+
+  def best_season(team_id)
+    win_count_by_season(team_id).max_by do |game_season, wins|
+      wins
+    end[0]
+  end
+
+  def season_home_losses(team_id)
+    losses = Hash.new(0.0)
+    home_games_grouped_by_season(team_id).each do |season, game|
+      game.each do |game|
+        if game.outcome == :away_win
+          losses[season] += 1.0
+        elsif game.outcome == :home_win
+          losses[season] -= 1.0
+        else
+          losses[season] += 0.5
+        end
+      end
+    end
+    losses
+  end
+
+  def season_away_losses(team_id)
+    losses = Hash.new(0.0)
+    away_games_grouped_by_season(team_id).each do |season, game|
+      game.each do |game|
+        if game.outcome == :home_win
+          losses[season] += 1.0
+        elsif game.outcome == :away_win
+          losses[season] -= 1.0
+        else
+          losses[season] += 0.5
+        end
+      end
+    end
+    losses
+  end
+
+  def loss_count_by_season(team_id)
+    accum = {}
+    season_away_losses(team_id).each do |season_away_id, away_loss|
+      season_home_losses(team_id).each do |season_home_id, home_loss|
+        if season_away_id == season_home_id
+          accum[season_away_id] = away_loss + home_loss
+        end
+      end
+    end
+    accum
+  end
+
+  def worst_season(team_id)
+    loss_count_by_season(team_id).max_by do |game_season, losses|
+      losses
+    end[0]
```

Comment:
Clever to tag on the [0] to the end block here, we should Implement this in season stats too